synapse/state/__init__.py

```diff
@@ -412,88 +436,76 @@ async def compute_event_context(
 
         delta_ids = {key: event.event_id}
 
-        state_group_after_event = (
-            await self._state_storage_controller.store_state_group(
-                event.event_id,
-                event.room_id,
+        if state_ids_before_event:
+            return UnpersistedEventContextForBatched(
+                storage_controller=self._storage_controllers,
+                state_group_before_event=state_group_before_event,
+                state_group=None,
+                state_delta_due_to_event=delta_ids,
                 prev_group=state_group_before_event,
                 delta_ids=delta_ids,
-                current_state_ids=None,
+                partial_state=partial_state,
+                state_map=state_ids_before_event,
             )
-        )
 
-        return EventContext.with_state(
-            storage=self._storage_controllers,
-            state_group=state_group_after_event,
-            state_group_before_event=state_group_before_event,
-            state_delta_due_to_event=delta_ids,
-            prev_group=state_group_before_event,
-            delta_ids=delta_ids,
-            partial_state=partial_state,
-        )
+        else:
+            return UnpersistedEventContext(
+                storage_controller=self._storage_controllers,
+                state_group_before_event=state_group_before_event,
+                state_group=None,
+                state_delta_due_to_event=delta_ids,
+                prev_group=state_group_before_event,
+                delta_ids=delta_ids,
+                partial_state=partial_state,
+            )
 
-    async def compute_event_context_for_batched(
+    async def compute_event_context(
         self,
         event: EventBase,
-        state_ids_before_event: StateMap[str],
-        current_state_group: int,
+        state_ids_before_event: Optional[StateMap[str]] = None,
+        partial_state: Optional[bool] = None,
+        entry: Optional[_StateCacheEntry] = None,
+        current_state_group: Optional[int] = None,
```

Comment:
Do we need any of these changes now?

Reply:
We can ditch `entry` for sure, `state_ids_before_event` and `partial_state` aren't actually changing what's there, i.e. it currently is https://github.com/matrix-org/synapse/blob/dd9e71dc7fa91b81adfaaf8669aaf7ee976ffcd7/synapse/state/__init__.py#L265. Removing the `current_state_group` probably makes sense too, as that will usually be used in conjunction with a batched event and so we'll want to separate out the persisting anyway using `calculate_context_info`.